pi/heartbeat.py

```diff
@@ -102,6 +103,33 @@ def ping(host):
     logging.info('returned: {}'.format(returncode))
     return returncode == 0
 
+def parse_link_quality_from_iwconfig_output(iwconfig_output_text):
+    lines = iwconfig_output_text.split('\n')
+    for i in range(0, len(lines)):
+        # note: hardcoded wireless network interface name
+        if ('wlan0' in lines[i]) and (i+5 < len(lines)):
+            # this should be something like 'Link Quality=69/70  Signal level=-41 dBm' which we then split
+            stats_strings = lines[i+5].split('  ')
+            for j in range(0, len(stats_strings)):
+                if 'Link Quality' in stats_strings[j]:
+                    quality_string = stats_strings[j].split('=')[1]
+                    # normalize the fraction since the denominator varies
+                    numerator = quality_string.split('/')[0]
+                    denominator = quality_string.split('/')[1]
+                    return float(numerator) / float(denominator)
+    return -1.0
+
+def log_wifi_link_quality():
+    try:
+        iwconfig_output_text = subprocess.check_output('iwconfig').decode('utf-8')
+        link_quality = parse_link_quality_from_iwconfig_output(iwconfig_output_text)
+        if link_quality < 0:
+            logging.warning("error parsing iwconfig output")
+        else:
+            logging.info('wlan0 link quality is %f', link_quality)
```

Comment:
Why not just output the whole `Link Quality ... Signal level.... ` line?

Reply:
I was thinking of wanting to graph this in the future, so it made sense to normalize the Link Quality fraction that `iwconfig` reports. I don't think the denominator is always the same, although it was always 70 for me, and sounds like it's the same for you on your machine.

The signal level is somewhat useful but for me it wasn't logging the noise level, and what we really want to know is the SNR, so the signal level alone is less useful. The `iwconfig` docs seem to indicate that the noise level will be logged, but I don't see it. That was my thinking anyway. And from my manual testing it seemed like the Link Quality metric was changing in response to real world things, so it seemed like a good place to start.

Reply:
Looking at this Stackoverflow answer (https://superuser.com/questions/866005/wireless-connection-link-quality-what-does-31-70-indicate), it looks like the denominator will always be 70. Whether to trust or not, is always the question.

For graphing, makes sense. I just figured since you were logging a string that you'd need to parse out the number from afterwards anyways, it seemed like less work to just use the string they already give you. Up to you!

Reply:
Hm, interesting post. The `iw` tool they mention yells not to 'screenscrape' it as the output format isn't stable apparently.

Fair point on the logging, I remember struggling to get the log files into the formats that I wanted them in for working with the Flic last seen values and wishing they were formatted in a more friendly way in the logs themselves. Hence my decision to implement this feature the way that I did.